lib/integrations/externalcloudaudit/bootstrap_test.go

```diff
@@ -0,0 +1,232 @@
+				AuditEventsLongTermURI: "s3://long-term-storage-bucket",
+				AthenaResultsURI:       "s3://transient-storage-bucket",
+				AthenaWorkgroup:        "teleport",
+				GlueDatabase:           "teleport",
+				GlueTable:              "audit-events",
+			},
+		},
+		{
+			desc:         "standard input bucket exists",
+			errExpected:  false,
+			region:       "us-west-2",
+			bucketExists: true,
+			eca: &ecatypes.ExternalCloudAuditSpec{
+				SessionsRecordingsURI:  "s3://long-term-storage-bucket",
+				AuditEventsLongTermURI: "s3://long-term-storage-bucket",
+				AthenaResultsURI:       "s3://transient-storage-bucket",
+				AthenaWorkgroup:        "teleport",
+				GlueDatabase:           "teleport",
+				GlueTable:              "audit-events",
+			},
+		},
+		{
+			desc:        "standard input table exists",
+			errExpected: false,
+			region:      "us-west-2",
+			tableExists: true,
+			eca: &ecatypes.ExternalCloudAuditSpec{
+				SessionsRecordingsURI:  "s3://long-term-storage-bucket",
+				AuditEventsLongTermURI: "s3://long-term-storage-bucket",
+				AthenaResultsURI:       "s3://transient-storage-bucket",
+				AthenaWorkgroup:        "teleport",
+				GlueDatabase:           "teleport",
+				GlueTable:              "audit-events",
+			},
+		},
+		{
+			desc:         "standard input database exists",
+			errExpected:  false,
+			region:       "us-west-2",
+			bucketExists: true,
+			eca: &ecatypes.ExternalCloudAuditSpec{
+				SessionsRecordingsURI:  "s3://long-term-storage-bucket",
+				AuditEventsLongTermURI: "s3://long-term-storage-bucket",
+				AthenaResultsURI:       "s3://transient-storage-bucket",
+				AthenaWorkgroup:        "teleport",
+				GlueDatabase:           "teleport",
+				GlueTable:              "audit-events",
+			},
+		},
+		{
+			desc:            "standard input workgroup exists",
+			errExpected:     false,
+			region:          "us-west-2",
+			workgroupExists: true,
+			eca: &ecatypes.ExternalCloudAuditSpec{
+				SessionsRecordingsURI:  "s3://long-term-storage-bucket",
+				AuditEventsLongTermURI: "s3://long-term-storage-bucket",
+				AthenaResultsURI:       "s3://transient-storage-bucket",
+				AthenaWorkgroup:        "teleport",
+				GlueDatabase:           "teleport",
+				GlueTable:              "audit-events",
+			},
+		},
+		{
+			desc:            "standard input all exist",
+			errExpected:     false,
+			region:          "us-west-2",
+			bucketExists:    true,
+			databaseExists:  true,
+			tableExists:     true,
+			workgroupExists: true,
+			eca: &ecatypes.ExternalCloudAuditSpec{
+				SessionsRecordingsURI:  "s3://long-term-storage-bucket",
+				AuditEventsLongTermURI: "s3://long-term-storage-bucket",
+				AthenaResultsURI:       "s3://transient-storage-bucket",
+				AthenaWorkgroup:        "teleport",
+				GlueDatabase:           "teleport",
+				GlueTable:              "audit-events",
+			},
+		},
+	}
+
+	for _, tc := range tt {
+		t.Run(tc.desc, func(t *testing.T) {
+			clt := &fakeBootstrapInfraClient{
```

Comment:
if you want to go one step further with this test, the fake client could actually store some representation of everything that got created, and you could then check that it's right

You could use this to write a test for idempotency by making the client actually return AlreadyExists errors if the thing was already created. I guess that's kind of what you're doing with the *Exists booleans but this way you could even call BootstrapInfra twice with semi-realistic results, and optionally make the first call fail halfway through with some other error

Reply:
Yea I definitely am looking for improvements on the testing side of things. I'll try to incorporate what you mentioned above and see how it turns out thanks.

Reply:
So I refactored slightly given your recommendations in [48cda44](https://github.com/gravitational/teleport/pull/33272/commits/48cda44499a48d7bcd25c1cd1e4957b306d70f2b). Let me know what you think and how I could potentially improve it further